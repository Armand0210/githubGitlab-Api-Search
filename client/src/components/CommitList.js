import React from "react";
import PropTypes from "prop-types";

//displaying an ordered list of props in the form of an array.
export function CommitList(props) {
  const commits = props.commits;
  //display message and date with unique key for each commit.
  const commitsList = commits.map((commit) => {
    return (
      <li key={commit.id} className="commit-bullet">
        {commit.message} <br></br>({commit.committed_date})
      </li>
    );
  });
  return <ol className="commits-list">{commitsList}</ol>;
}
//required props
CommitList.propTypes = {
  commits: PropTypes.array.isRequired,
};
