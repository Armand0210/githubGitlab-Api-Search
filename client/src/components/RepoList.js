import React from "react";
import PropTypes from "prop-types";

//Displays props as ordered lists
export function RepoList(props) {
  //fetches list of repos from given props
  const repos = props.repos;
  //when butten is clicked invokes displayrepo function(prop)
  const handleClick = (e) => {
    props.displayRepo(props.user.source, props.user.login, e.target.value);
  };
  //Display name and assign unique id/key.
  const reposList = repos.map((repo) => {
    return (
      <li key={repo.id}>
        <div className="repo-bullet">
          <span>{repo.name}</span>
          <button
            //assigning 2 values, the server IP uses these two
            value={[repo.name, repo.id]}
            onClick={handleClick}
          >
            View repo
          </button>
        </div>
      </li>
    );
  });
  return <ol className="repo-list">{reposList}</ol>;
}

//required props
RepoList.propTypes = {
  repos: PropTypes.array.isRequired,
  user: PropTypes.object.isRequired,
  displayRepo: PropTypes.func.isRequired,
};
