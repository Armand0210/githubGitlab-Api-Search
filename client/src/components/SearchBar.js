import React from "react";
import PropTypes from "prop-types";

//takes prop in form of search function and displays form.
//if input is intered it updates its own state when form is submitted.
export class SearchBar extends React.Component {
  constructor(props) {
    super(props);
    this.state = { searchString: "" };
    //Binding two functions below.
    this.handleUserInput = this.handleUserInput.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  //when input is detected inside input field, Update state to reflect new values.
  handleUserInput(e) {
    this.setState({ searchString: e.target.value });
  }
  //When form is submitted - begin search
  handleSubmit(e) {
    //prevent refresh
    e.preventDefault();
    this.props.handleSearch(this.state.searchString);
    //clearing input Field
    this.setState({ searchString: "" });
  }
  render() {
    return (
      <div>
        <form onSubmit={this.handleSubmit}>
          <input
            type="text"
            value={this.state.searchString}
            onChange={this.handleUserInput}
            className="search-input"
            required
          ></input>
          <button type="submit">Search</button>
        </form>
      </div>
    );
  }
}

//required prop
SearchBar.propTypes = {
  handleSearch: PropTypes.func.isRequired,
};
