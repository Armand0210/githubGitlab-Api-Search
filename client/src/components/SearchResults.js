import React from "react";
import PropTypes from "prop-types";

//this function takes in prop in the form of an array of result objects.
export function SearchResults(props) {
  //fetches list of repos from given props
  const results = props.results;
  //triggered when button is clicked, invokes displayUser function (prop)
  const handleClick = (e) => {
    props.displayUser(e.target.value);
  };
  //per search result display name,source and assign key to it
  const resultsList = results.map((result) => {
    return (
      <li key={result.id} className="result-bullet">
        <p>
          <strong>Username: </strong>
          {result.login} <br></br>
          <strong>Source: </strong>
          {result.source} <br></br>
        </p>
        <button
          //2 values needed by api
          value={[result.source, result.login]}
          onClick={handleClick}
        >
          View user
        </button>
      </li>
    );
  });
  return <ul className="results-list">{resultsList}</ul>;
}

//required props
SearchResults.propTypes = {
  results: PropTypes.array.isRequired,
  displayUser: PropTypes.func.isRequired,
};
