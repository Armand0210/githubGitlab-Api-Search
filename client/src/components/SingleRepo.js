import React from "react";
import { CommitList } from "./CommitList";
import PropTypes from "prop-types";
import { getData } from "../utils/getData.js";

//this function is used to take a prop in the form of a repo project.
export function SingleRepo(props) {
  //top 5 commit info
  const [commits, setCommits] = React.useState(null);
  //when component is rendered the function will render and fetch the list of commits.
  React.useEffect(() => {
    const fetchData = async () => {
      //constructing server end point
      const url = `/api/commits?source=${props.repo.source}&user=${props.repo.owner}&reponame=${props.repo.name}&repoid=${props.repo.id}`;
      //fetching the data and storing it in result variable.
      const result = await getData(url);
      //updating the state to store the variable contents.
      setCommits(result);
    };
    //calling function for fetch
    fetchData();
  }, [props.repo]);
  return (
    <div className="single-repo">
      <h2>{props.repo.name}</h2>
      <div className="abc">
        <p>
          {props.repo.description}
          <br></br>
          <br></br>
          <strong>Created Date: </strong>
          {props.repo.created_at}
          <br></br>
          <strong>Last Updates Date: </strong>
          {props.repo.updated_at}
          <br></br>
        </p>
        <h3>Last 5 commits:</h3>
        {!commits ? (
          <div className="placeholder">Loading...</div>
        ) : commits.length === 0 ? (
          <div className="placeholder">No commits yet</div>
        ) : (
          <CommitList commits={commits}></CommitList>
        )}
      </div>
      <div className="abc">
        <button onClick={props.backToUser}>Back to profile</button>
      </div>
    </div>
  );
}

//required props
SingleRepo.propTypes = {
  repo: PropTypes.object.isRequired,
  backToUser: PropTypes.func.isRequired,
};
