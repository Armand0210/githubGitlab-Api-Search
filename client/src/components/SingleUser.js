import React from "react";
import { RepoList } from "./RepoList";
import PropTypes from "prop-types";
import { getData } from "../utils/getData.js";

//takes prop in the form of a user object and displays the details.
export function SingleUser(props) {
  //contain top 5 repos
  const [repos, setRepos] = React.useState(null);
  //on render the function will run and fetch list of repos
  React.useEffect(() => {
    const fetchData = async () => {
      const url = `/api/repos?source=${props.user.source}&user=${props.user.login}`;
      const result = await getData(url);
      setRepos(result);
    };
    fetchData();
    //running every time users info changes
  }, [props.user]);

  return (
    <div className="single-user">
      <h2>{props.user.login}</h2>
      <div className="user-details">
        <img src={props.user.avatar} alt="user avatar" className="avatar"></img>
        <p>
          <a href={props.user.url}>{props.user.url}</a>
        </p>
        <h3>Last 5 repositories:</h3>
        {!repos ? (
          <div className="placeholder">Fetching last 5 repositories</div>
        ) : repos.length === 0 ? (
          <div className="placeholder">No repositories</div>
        ) : (
          <RepoList
            user={props.user}
            repos={repos}
            displayRepo={props.displayRepo}
          ></RepoList>
        )}
      </div>
      <div className="abc">
        <button onClick={props.backToSearch}>Back to results</button>
      </div>
    </div>
  );
}

//required props
SingleUser.propTypes = {
  user: PropTypes.object.isRequired,
  backToSearch: PropTypes.func.isRequired,
  displayRepo: PropTypes.func.isRequired,
};
