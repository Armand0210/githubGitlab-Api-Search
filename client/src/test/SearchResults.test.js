import React from "react";
import { SearchResults } from "../components/SearchResults";
import renderer from "react-test-renderer";
//Tested a random user - test if components renders correctly
test("SearchResults shows results when a search is complete", () => {
  let resultSet = [
    {
      id: 737241,
      name: null,
      login: "Ram-N",
      avatar: "https://avatars.githubusercontent.com/u/737241?v=4",
      url: "https://api.github.com/users/Ram-N",
      source: "Github",
    },
    {
      id: 243,
      name: "inkscape",
      login: "inkscape",
      avatar:
        "https://gitlab.com/uploads/-/system/project/avatar/15192091/inkscape.png?width=64",
      url: "https://gitlab.com/inkscape",
      source: "Gitlab",
    },
  ];
  const tree = renderer
    .create(
      <SearchResults
        results={resultSet}
        displayUser={function displayUser() {}}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
