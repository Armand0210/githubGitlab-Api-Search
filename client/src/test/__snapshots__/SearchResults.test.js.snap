// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResults shows results when a search is complete 1`] = `
<ul
  className="results-list"
>
  <li
    className="result-bullet"
  >
    <p>
      <strong>
        Username: 
      </strong>
      Ram-N
       
      <br />
      <strong>
        Source: 
      </strong>
      Github
       
      <br />
    </p>
    <button
      onClick={[Function]}
      value={
        Array [
          "Github",
          "Ram-N",
        ]
      }
    >
      View user
    </button>
  </li>
  <li
    className="result-bullet"
  >
    <p>
      <strong>
        Username: 
      </strong>
      inkscape
       
      <br />
      <strong>
        Source: 
      </strong>
      Gitlab
       
      <br />
    </p>
    <button
      onClick={[Function]}
      value={
        Array [
          "Gitlab",
          "inkscape",
        ]
      }
    >
      View user
    </button>
  </li>
</ul>
`;
